### PR TITLE
[Merged by Bors] - Make grpc servers' start methods be synchronous

### LINKED ISSUE #4861: GRPC servers unnecessarily do startup asynchronously
## Description
`grpcserver.Server.Start` and `grpcserver.JSONHTTPServer.Start` return a channel upon which the caller can wait to be notified of the service having completed startup. It doesn't appear necessary and can easily be overlooked by callers.

Instead these methods should be synchronous, so that when they return the caller can safely connect to the server.
## Affected code
`api/grpcserver` package.
This issue appears in commit hash: c7a7468f3bc2ed2a159c2c81e90297deb5757efc

### PR DESCRIPTION
**Pull request by piersy**
## Motivation
Closes #4861 
## Changes

Updates grpc server startup code to execute synchronously, returning only after the server has started listening.

## Test Plan
Existing tests should pass

## DevOps Notes

- [x] This PR does not require configuration changes (e.g., environment variables, GitHub secrets, VM resources)
- [x] This PR does not affect public APIs
- [x] This PR does not rely on a new version of external services (PoET, elasticsearch, etc.)
- [x] This PR does not make changes to log messages (which monitoring infrastructure may rely on)

### CODE REVIEW

File: api/grpcserver/grpcserver_test.go [hunk 1]
```diff
@@ -491,20 +491,14 @@ func launchServer(tb testing.TB, cfg Config, services ...ServiceAPI) func() {
...
 	// start gRPC and json servers
-	grpcStarted := grpcService.Start()
-	jsonStarted := jsonService.StartService(context.Background(), services...)
+	err := grpcService.Start()
+	require.NoError(tb, err)
+	err = jsonService.StartService(context.Background(), services...)
+	require.NoError(tb, err)
 
 	timer := time.NewTimer(3 * time.Second)
 	defer timer.Stop()
```
Comment (fasmat) [inline api/grpcserver/grpcserver_test.go]: I think this is not needed any more 🙂

File: cmd/bootstrapper/generator_test.go [hunk 2]
```diff
@@ -139,20 +139,14 @@ func launchServer(tb testing.TB, cdb *datastore.CachedDB) func() {
...
 	// start gRPC and json servers
-	grpcStarted := grpcService.Start()
-	jsonStarted := jsonService.StartService(context.Background(), s)
+	err := grpcService.Start()
+	require.NoError(tb, err)
+	err = jsonService.StartService(context.Background(), s)
+	require.NoError(tb, err)
 
 	timer := time.NewTimer(3 * time.Second)
 	defer timer.Stop()
```
Comment (fasmat) [inline cmd/bootstrapper/generator_test.go]: Same here, I don't think this is needed any more.

**Review by fasmat — APPROVED** [review]

File: api/grpcserver/grpc.go [hunk 3]
```diff
@@ -44,35 +46,27 @@ func (s *Server) Start() <-chan struct{} {
...
+		}
+		return nil
+	})
+	return nil
 }
 
 // Close stops the server.
 func (s *Server) Close() error {
 	s.logger.Info("stopping the grpc server")
 	s.GrpcServer.Stop()
```
Comment (poszu) [inline api/grpcserver/grpc.go:68]: Consider a graceful stop to allow pending GRPC calls to finish:
```suggestion
	s.GrpcServer.GracefulStop()
```

Comment (piersy) [reply api/grpcserver/grpc.go:68]: I don't think just calling GracefulStop would work well with our streaming APIs, I think if you had a stream open the server would never shutdown. The approach to handling this, according to this [ticket](https://github.com/grpc/grpc-go/issues/2888), is to first call GracefulStop, wait some time and then call Stop. I can add a ticket for it, it seems to be out of scope for this PR.

**Comment by piersy** [comment]
bors merge

File: api/grpcserver/grpc.go (hunk 3 above)
Comment (poszu) [reply api/grpcserver/grpc.go:68]: Thanks, didn't know that! Do you think it's worth creating an issue to implement the proposed graceful shutdown later?

Comment (piersy) [reply api/grpcserver/grpc.go:68]: Sure will do!

Comment (piersy) [reply api/grpcserver/grpc.go:68]: https://github.com/spacemeshos/go-spacemesh/issues/4867

**Comment by piersy** [comment]
bors cancel

**Comment by piersy** [comment]
bors merge

**Comment by piersy** [comment]
bors merge

**Pull request closed without merging**